src/parser.cpp

```diff
@@ -1006,7 +1006,10 @@ namespace ada::parser {
             // and append the result to url’s path.
             if (pointer != pointer_end) {
               if (character_sets::bit_at(character_sets::C0_CONTROL_PERCENT_ENCODE, *pointer)) {
-                url.path += character_sets::hex + *pointer * 4;
+                // We cast to an unsigned 8-bit integer because
+                // *pointer is of type 'char' which may be signed or unsigned.
+                // A negative index access in 'character_sets::hex' is unsafe.
+                url.path += character_sets::hex[uint8_t(*pointer) * 4];
```

Comment:
This is also used inside unicode.cpp. We should update it too.

Reply:
I have not seen other such problems.